File: front/src/lib/blindbox.ts
import axios from 'axios';
import apiClient from '@/lib/api';
import { API_URLS } from '@/conf/env';
import { Product } from './product';

export interface BlindBox {
  id: string;
  name: string;
  description: string;
  discountCoefficient: number;
  category: string;
  imageUrl: string;
  status: string;
  expirationTime: string;
  donationAmount: number;
  productCount: number;
  createdAt: string;
  creatorId?: string;
}

export interface BlindBoxDetail extends BlindBox {
  products: Product[];
}

// Define ProductDTO directly here based on backend response
export interface ProductDTO {
  id: string;
  name: string;
  description?: string;
  price: number;
  category: string;
  imageUrl?: string;
  status: string;
  productionDate: string;
  shelfLifeHours: number;
  createdAt: string;
}

// 用于创建/更新盲盒的数据结构
export interface BlindBoxInputData {
  name: string;
  description?: string;
  discountCoefficient: number;
  category: string;
  imageURL?: string;
  donationAmount?: number;
  expirationTime?: string;
}

export interface BlindBoxListOptions {
  page?: number;
  size?: number;
  category?: string;
  status?: string;
  keyword?: string;
  sortBy?: string;
  order?: string;
  minPrice?: number;
  maxPrice?: number;
}

export interface BlindBoxListResponse {
  boxes: BlindBox[];
  total: number;
  page: number;
  size: number;
}

// 购买盲盒响应
export interface PurchaseResponse {
  message: string;
  order_id: string;
  box_id: string;
  price: number;
  status: string;
}

// 开启盲盒响应
export interface OpenBoxResponse {
  message: string;
  opening_id: string;
  box_id: string;
  product_id: string;
  product_name: string;
  product_info: ProductDTO; // Use the locally defined DTO
  opened_at: string;
}

export const blindboxService = {
  // 获取盲盒列表
  async listBlindBoxes(options: BlindBoxListOptions): Promise<BlindBoxListResponse> {
    const params = new URLSearchParams();
    if (options.page) params.append('page', options.page.toString());
    if (options.size) params.append('size', options.size.toString());
    if (options.category) params.append('category', options.category);
    if (options.status) params.append('status', options.status);
    if (options.keyword) params.append('keyword', options.keyword);
    if (options.sortBy) params.append('sort_by', options.sortBy);
    if (options.order) params.append('order', options.order);
    if (options.minPrice) params.append('min_price', options.minPrice.toString());
    if (options.maxPrice) params.append('max_price', options.maxPrice.toString());

    const response = await axios.get(`${API_URLS.BOX.LIST}?${params.toString()}`);
    return response.data;
  },

  // 获取盲盒详情
  async getBlindBox(id: string): Promise<BlindBoxDetail> {
    const response = await axios.get(API_URLS.BOX.DETAIL(id));
    return response.data;
  },

  // 创建盲盒
  async createBlindBox(data: BlindBoxInputData): Promise<BlindBox> {
    const response = await axios.post(API_URLS.BOX.CREATE, data);
    // Backend returns { message, box_id }, let's assume we need the box details after creation for consistency.
    // This might require an additional getBlindBox call or backend changes.
    // For now, returning the simple response to match backend.
    // return response.data; 
    // To make it work with front-end expectations (needs a BlindBox object)
    // We return a partial object, real data needs refresh or backend change.
    return { id: response.data.box_id, ...data } as BlindBox; 
  },

  // 更新盲盒
  async updateBlindBox(id: string, data: Partial<BlindBoxInputData>): Promise<BlindBox> {
    await axios.put(API_URLS.BOX.UPDATE(id), data);
    // Backend returns { message, box_id }, we need to get the updated box details
    const updatedBox = await this.getBlindBox(id);
    return updatedBox;
  },

  // 删除盲盒
  async deleteBlindBox(id: string): Promise<void> {
    await axios.delete(API_URLS.BOX.DELETE(id));
  },

  // 购买盲盒
  async purchaseBlindBox(id: string): Promise<PurchaseResponse> {
    const response = await axios.post(API_URLS.BOX.PURCHASE(id));
    return response.data;
  },

  // 开启盲盒
  async openBlindBox(id: string): Promise<OpenBoxResponse> {
    const response = await axios.post(API_URLS.BOX.OPEN(id));
    return response.data;
  },

  // 获取盲盒开启历史
  async getBlindBoxOpeningHistory(id: string): Promise<any[]> {
    const response = await axios.get(API_URLS.BOX.HISTORY(id));
    return response.data.openings;
  },

  // 添加产品到盲盒
  async addProductToBlindBox(productId: string, boxId: string): Promise<void> {
    await axios.post(API_URLS.PRODUCT.ADD_TO_BOX(productId), { blind_box_id: boxId });
  },

  // 从盲盒移除产品
  async removeProductFromBlindBox(productId: string): Promise<void> {
    await axios.post(API_URLS.PRODUCT.REMOVE_FROM_BOX(productId));
  },

  // 批量添加产品到盲盒
  async batchAddProductsToBlindBox(productIds: string[], boxId: string): Promise<void> {
    await axios.post(API_URLS.PRODUCT.BATCH_ADD_TO_BOX, {
      product_ids: productIds,
      blind_box_id: boxId,
    });
  },

  // 上传盲盒封面图片
  async uploadBlindboxCover(file: File): Promise<string> {
    const formData = new FormData();
    formData.append('image', file);

    const response = await apiClient.post(API_URLS.UPLOAD.WITH_TYPE('box'), formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });

    // 约定后端返回 { code, data: { url } }
    return response.data.data?.url || '';
  },
};

File: internal/api/rest/handler/blindbox.go
package handler

import (
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"go.uber.org/zap"

	"FreshBox/internal/model"
	"FreshBox/internal/service"
)

// BlindBoxHandler 盲盒处理器
type BlindBoxHandler struct {
	blindBoxService *service.BlindBoxService
	productService  *service.ProductService
	logger          *zap.Logger
}

// NewBlindBoxHandler 创建盲盒处理器
func NewBlindBoxHandler(blindBoxService *service.BlindBoxService, productService *service.ProductService) *BlindBoxHandler {
	logger, _ := zap.NewDevelopment()
	return &BlindBoxHandler{
		blindBoxService: blindBoxService,
		productService:  productService,
		logger:          logger,
	}
}

// CreateBlindBox 创建盲盒
func (h *BlindBoxHandler) CreateBlindBox(c *gin.Context) {
	var box model.BlindBox
	if err := c.ShouldBindJSON(&box); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "参数错误", "details": err.Error()})
		return
	}

	// 设置创建者ID（从认证中获取）
	userID := c.GetString("user_id")
	if userID == "" {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "未授权"})
		return
	}
	box.CreatorID = userID

	// 设置过期时间（如果未提供）
	if box.ExpirationTime.IsZero() {
		box.ExpirationTime = time.Now().Add(72 * time.Hour) // 默认3天后过期
	}

	err := h.blindBoxService.CreateBlindBox(c.Request.Context(), &box, userID)
	if err != nil {
		h.logger.Error("创建盲盒失败", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "创建盲盒失败", "details": err.Error()})
		return
	}

	c.JSON(http.StatusCreated, gin.H{
		"message": "创建盲盒成功",
		"box_id":  box.ID,
	})
}

// GetBlindBox 获取盲盒详情
func (h *BlindBoxHandler) GetBlindBox(c *gin.Context) {
	id := c.Param("id")
	if id == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "盲盒ID不能为空"})
		return
	}

	box, err := h.blindBoxService.GetBlindBox(c.Request.Context(), id)
	if err != nil {
		h.logger.Error("获取盲盒失败", zap.Error(err), zap.String("box_id", id))
		c.JSON(http.StatusNotFound, gin.H{"error": "获取盲盒失败", "details": err.Error()})
		return
	}

	// 获取盲盒内的商品
	products, err := h.productService.GetProductsByBlindBox(c.Request.Context(), id)
	if err != nil {
		h.logger.Error("获取盲盒商品失败", zap.Error(err), zap.String("box_id", id))
	}

	// 构建响应DTO
	boxDTO := model.BlindBoxDetailDTO{
		BlindBoxDTO: model.BlindBoxDTO{
			ID:                  box.ID,
			Name:                box.Name,
			Description:         box.Description,
			DiscountCoefficient: box.DiscountCoefficient,
			Category:            box.Category,
			ImageURL:            box.ImageURL,
			Status:              box.Status,
			ExpirationTime:      service.FormatTime(box.ExpirationTime),
			DonationAmount:      box.DonationAmount,
			ProductCount:        len(products),
			CreatedAt:           service.FormatTime(box.CreatedAt),
		},
		Products: []model.ProductDTO{},
	}

	// 转换商品到DTO
	for _, p := range products {
		boxDTO.Products = append(boxDTO.Products, model.ProductDTO{
			ID:             p.ID,
			Name:           p.Name,
			Description:    p.Description,
			Price:          p.Price,
			Category:       p.Category,
			ImageURL:       p.ImageURL,
			Status:         p.Status,
			ProductionDate: service.FormatTime(p.ProductionDate),
			ShelfLifeHours: p.ShelfLifeHours,
			CreatedAt:      service.FormatTime(p.CreatedAt),
		})
	}

	c.JSON(http.StatusOK, boxDTO)
}

// ListBlindBoxes 列出盲盒
func (h *BlindBoxHandler) ListBlindBoxes(c *gin.Context) {
	// 解析查询参数
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	size, _ := strconv.Atoi(c.DefaultQuery("size", "10"))
	category := c.Query("category")
	status := c.Query("status")
	keyword := c.Query("keyword")
	sortBy := c.Query("sort_by")
	order := c.Query("order")
	minPrice, _ := strconv.ParseFloat(c.Query("min_price"), 64)
	maxPrice, _ := strconv.ParseFloat(c.Query("max_price"), 64)

	// 构建查询选项
	opts := model.BlindBoxListOptions{
		Page:     page,
		Size:     size,
		Category: category,
		Status:   status,
		Keyword:  keyword,
		SortBy:   sortBy,
		Order:    order,
		MinPrice: minPrice,
		MaxPrice: maxPrice,
	}

	// 查询盲盒列表
	boxes, total, err := h.blindBoxService.ListBlindBoxes(c.Request.Context(), opts)
	if err != nil {
		h.logger.Error("查询盲盒列表失败", zap.Error(err))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "查询盲盒列表失败", "details": err.Error()})
		return
	}

	// 构建响应DTO
	var boxDTOs []model.BlindBoxDTO
	for _, box := range boxes {
		// 获取盲盒内商品数量
		productCount, _ := h.productService.CountProductsByBlindBox(c.Request.Context(), box.ID)

		boxDTOs = append(boxDTOs, model.BlindBoxDTO{
			ID:                  box.ID,
			Name:                box.Name,
			Description:         box.Description,
			DiscountCoefficient: box.DiscountCoefficient,
			Category:            box.Category,
			ImageURL:            box.ImageURL,
			Status:              box.Status,
			ExpirationTime:      service.FormatTime(box.ExpirationTime),
			DonationAmount:      box.DonationAmount,
			ProductCount:        int(productCount),
			CreatedAt:           service.FormatTime(box.CreatedAt),
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"boxes": boxDTOs,
		"total": total,
		"page":  page,
		"size":  size,
	})
}

// UpdateBlindBox 更新盲盒
func (h *BlindBoxHandler) UpdateBlindBox(c *gin.Context) {
	id := c.Param("id")
	if id == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "盲盒ID不能为空"})
		return
	}

	var box model.BlindBox
	if err := c.ShouldBindJSON(&box); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "参数错误", "details": err.Error()})
		return
	}

	// 设置ID
	box.ID = id

	// 获取用户ID
	userID := c.GetString("user_id")
	if userID == "" {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "未授权"})
		return
	}

	err := h.blindBoxService.UpdateBlindBox(c.Request.Context(), id, &box, userID)
	if err != nil {
		h.logger.Error("更新盲盒失败", zap.Error(err), zap.String("box_id", id))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "更新盲盒失败", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "更新盲盒成功",
		"box_id":  id,
	})
}

// DeleteBlindBox 删除盲盒
func (h *BlindBoxHandler) DeleteBlindBox(c *gin.Context) {
	id := c.Param("id")
	if id == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "盲盒ID不能为空"})
		return
	}

	// 获取用户ID
	userID := c.GetString("user_id")
	if userID == "" {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "未授权"})
		return
	}

	err := h.blindBoxService.DeleteBlindBox(c.Request.Context(), id, userID)
	if err != nil {
		h.logger.Error("删除盲盒失败", zap.Error(err), zap.String("box_id", id))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "删除盲盒失败", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "删除盲盒成功",
		"box_id":  id,
	})
}

// PurchaseBlindBox 购买盲盒
func (h *BlindBoxHandler) PurchaseBlindBox(c *gin.Context) {
	id := c.Param("id")
	if id == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "盲盒ID不能为空"})
		return
	}

	// 获取用户ID
	userID := c.GetString("user_id")
	if userID == "" {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "未授权"})
		return
	}

	order, err := h.blindBoxService.PurchaseBlindBox(c.Request.Context(), id, userID)
	if err != nil {
		h.logger.Error("购买盲盒失败", zap.Error(err), zap.String("box_id", id))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "购买盲盒失败", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message":  "购买盲盒成功",
		"order_id": order.ID,
		"box_id":   id,
		"price":    order.Price,
		"status":   order.Status,
	})
}

// OpenBlindBox 开启盲盒
func (h *BlindBoxHandler) OpenBlindBox(c *gin.Context) {
	id := c.Param("id")
	if id == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "盲盒ID不能为空"})
		return
	}

	// 获取用户ID
	userID := c.GetString("user_id")
	if userID == "" {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "未授权"})
		return
	}

	opening, err := h.blindBoxService.OpenBlindBox(c.Request.Context(), id, userID)
	if err != nil {
		h.logger.Error("开启盲盒失败", zap.Error(err), zap.String("box_id", id))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "开启盲盒失败", "details": err.Error()})
		return
	}

	// 获取获得的商品详情
	product, err := h.productService.GetProduct(c.Request.Context(), opening.ObtainedProductID)
	if err != nil {
		h.logger.Error("获取商品详情失败", zap.Error(err), zap.String("product_id", opening.ObtainedProductID))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "获取商品详情失败", "details": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message":      "开启盲盒成功",
		"opening_id":   opening.ID,
		"box_id":       id,
		"product_id":   product.ID,
		"product_name": product.Name,
		"product_info": model.ProductDTO{
			ID:             product.ID,
			Name:           product.Name,
			Description:    product.Description,
			Price:          product.Price,
			Category:       product.Category,
			ImageURL:       product.ImageURL,
			Status:         product.Status,
			ProductionDate: service.FormatTime(product.ProductionDate),
			ShelfLifeHours: product.ShelfLifeHours,
			CreatedAt:      service.FormatTime(product.CreatedAt),
		},
		"opened_at": service.FormatTime(opening.OpenedAt),
	})
}

// GetBlindBoxOpeningHistory 获取盲盒开启历史
func (h *BlindBoxHandler) GetBlindBoxOpeningHistory(c *gin.Context) {
	id := c.Param("id")
	if id == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "盲盒ID不能为空"})
		return
	}

	openings, err := h.blindBoxService.GetBlindBoxOpeningHistory(c.Request.Context(), id)
	if err != nil {
		h.logger.Error("获取盲盒开启历史失败", zap.Error(err), zap.String("box_id", id))
		c.JSON(http.StatusInternalServerError, gin.H{"error": "获取盲盒开启历史失败", "details": err.Error()})
		return
	}

	var openingDTOs []gin.H
	for _, opening := range openings {
		// 获取用户信息
		// 此处应该调用userService获取用户信息，这里简化处理
		openingDTOs = append(openingDTOs, gin.H{
			"opening_id": opening.ID,
			"user_id":    opening.UserID,
			"box_id":     opening.BlindBoxID,
			"product_id": opening.ObtainedProductID,
			"opened_at":  service.FormatTime(opening.OpenedAt),
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"openings": openingDTOs,
		"box_id":   id,
		"total":    len(openingDTOs),
	})
}
